docs/website/contents/for-developers/HardWonWisdom.md

```diff
@@ -348,3 +348,77 @@ The key corresponding parts of the implementation are as follows.
 - The interpretation only depends on the eras' start and end bounds, not on the safe zone.
   The safe zone is instead used in an preceding step to safely determine the least possible value of the first unknown end bound.
   This determination is implemented by the `ouroboros-consensus:Ouroboros.Consensus.HardFork.Combinator.State.reconstructSummaryLedger` function.
+
+## The hard fork combinator interface
+
+To run the Consensus Layer on a block type `blk` requires the following (whether or not it's a `HardForkBlock`).
+
+- Values for various tracers, ChainDB options, etc, as required by `Ouroboros.Consensus.Node.run`.
+- A `ProtocolInfo blk` value, which is `TopLevelConfig blk` as well as initial state of the ledger and protocol.
+- Instances of `ConsensusProtocol (BlockProtocol blk)`, `LedgerSupportsProtocol blk`, and so on (see the `RunNode blk` class).
+
+Suppose there are two blocks types `F` and `G` for which the above is already available.
+To run the Consensus Layer instead on the hard fork combinator that sequences `F` into `G` requires the following additional information about `F` and `G`.
+
+- A value of `EraParams` for `F` and for `G` (ie `Shape [F, G]`), which decomposes as follows.
+    - The duration of every slot in the epoch as a [`NominalDiffTime`](https://hackage.haskell.org/package/time-1.12.2/docs/Data-Time-Clock.html#t:NominalDiffTime).
+    - The duration of every epoch in the era as a number of slots.
+    - The era's _safe zone_, which depends on the `singleEraTransition` function; see below.
+        - If the era's safe zone is `d`, then `singleEraTransition` cannot return `Just e` for the ledger state resulting from some block `x` (possibly the "genesis block") (possibly from an earlier era, and not even necessarily the previous era) unless there are at least `d` slots between `x` and `e`, ie `slot(x) + d < firstSlot(e)`.
```

Comment:
> If the era's safe zone is `d`, then `singleEraTransition` cannot return `Just e` for the ledger state resulting from some block `x` (possibly the "genesis block") (possibly from an earlier era, and not even necessarily the previous era) unless there are at least `d` slots between `x` and `e`, ie `slot(x) + d < firstSlot(e)`.

Hmm, isn't it completely fine for `singleEraTransition` to return `Just e` within one safe zone before `e` starts as long as it was also known before that (ie a ledger state with tip before one safe zone before `e` started also caused `singleEraTransition` to return `Just e`)? Eg our implementation for eg Shelley will happily do that.

I guess one could fix that by talking about the *first time* `singleEraTransition` reports `Just e` (this must happen at least one stability window before `e` starts).

Maybe the definition from the report could be repeated here again (to keep it more high-level):
> **Definition 17.3** (Safe zone)**.** A safe zone is a period of time ahead of a ledger’s tip in which an era transition is guaranteed not to occur if it is not yet known.

Maybe we could try to translate this into a law for `singleEraTransition`, but it would have to talk about rollbacks, which you already do further below.

Reply:
OK. I reorganized and I think it's all technically correct now.

Reply:
Ah ha! I fixed one problem, but not the whole thing. I'm on it now.